### ISSUE TITLE
http transport appears to block/timeout

### ISSUE THREAD

**Issue by schoolcoder** [meetecho/janus-gateway#2096]
We are using Janus http and websocket transports. 

The websocket transport is used from our backend, and works without problem.

The http transport is used from our browser client (using janus.js). It also works troublefree, but then, suddenly, the janus server quits responding. All connect requests timeout. We have debug level at 5;  there are no error messages in the janus log file.

After http starts failing, websocket continues to work.

We are using videoroom plugin, but problem does not seem directly related to the plugin. What fails in http is session creation, before we even get to attaching the videoroom handle.

Also, it isn't a bug in janus.js. In our backend, we implemented a health probe that also uses the same http transport as our browser client. The health probe fails (times out) just as the browser does.

We are using apache as a front end and proxying to /janus. However, apache is not reporting any problems. When the failure occurs, restarting apache has no effect, but restarting janus fixes it immediately.

I have not confirmed yet whether the 'info' command continues to work when session creation fails.

It feels like something (a dangling mutex?) is blocking the http transport. I took a look at the http transport code, but didn't see anything.

Our server is very, very lightly loaded--definitely not a load issue. The sequence of events is:

1. Janus videoroom has no rooms.
2. User wants to enter room A.
3. Backend creates (via websocket) room A on janus. No problem.
4. User tries to enter room A (using janus.js via http). Times out.

This problem occurs infrequently. One janus server has been running for 2 weeks without any problem. On another server, it happened in less than 1 day.

Could our dual use of websockets and http be the cause of the problem? Any advice is appreciated.

**Comment by schoolcoder** [comment]
I took a look at janus_http.c and janus_websockets.c

In both files, there are several global variables that (I think) should be declared 'static'.

In janus_http.c:
```
static const char *keepalive_id = "keepalive";
static GHashTable *sessions = NULL;
static janus_mutex sessions_mutex = JANUS_MUTEX_INITIALIZER;
static GList *janus_http_access_list = NULL, *janus_http_admin_access_list = NULL;
static janus_mutex access_list_mutex;
```

In janus_websockets.c:
```
static GList *janus_websockets_access_list = NULL, *janus_websockets_admin_access_list = NULL;
static janus_mutex access_list_mutex;
```

The only common variable is `access_list_mutex`. That means a duplicate global symbol at runtime. In Ubuntu 18.04, I tried a simple test and confirmed that, in the case of a duplicate global symbol, only the first instance (based on link order) is accessible. So, for example, at runtime, the janus_http code's references to `access_list_mutex` could be actually accessing the variable in janus_websockets. Or vice versa.

Could this be the cause? Then again, the duplicate symbol isn't optimal, but it doesn't look like it would cause any harm...

**Comment by atoppi** [comment]
Hi @schoolcoder,
we have fixed many issues in the HTTP transport in the last days (including ones similar to your report), so please test with master and report back in case of problems.

As of your findings about the `access_list_mutex` you're right, the linker defines them as global symbols so probably at runtime we're using the same mutex for both transports.

(uppercase symbol type means GLOBAL)
```
$ nm -a /usr/local/janus/lib/janus/transports/libjanus_http.so | grep access_list_mutex
0000000000297d68 B access_list_mutex

$ nm -a /usr/local/janus/lib/janus/transports/libjanus_websockets.so | grep access_list_mutex
000000000025c3e8 B access_list_mutex
```

Anyway we're just sharing the same mutex when accessing the transport ACLs, so this should not be a big issue.

We'll fix this and other _forgotten_ static declarations, thanks for spotting it.

**Comment by lminiero** [comment]
Static stuff fixed here: https://github.com/meetecho/janus-gateway/commit/4e944fe3f73de596ae7b97db7b2b9d58ad420583
Please test and check if it helps. That said, without info on the revision you're using and debugging info (e.g., lock debugging), nothing else we can do.

**Comment by Haesmen** [comment]
I confirm this problem on release 0.9.3 and the master branch, on 0.9.1 http transport works without problems

**Comment by lminiero** [comment]
Confirming without feedback or info is exactly what we DON'T need.

**Comment by schoolcoder** [comment]
We will pull the latest, turn on lock debugging and debug level 7, and collect more info. Thanks for the static fix, too.

When the problem occurs, there are absolutely no messages in the janus log (at debug level 5, lock debugging off). Ordinarily, we see "`Got a Janus API request from janus.transport.http`", but nothing when http is unresponsive.

**Comment by lminiero** [comment]
No debug level 7, please: that's just noise. 5 is way more than enough.

**Comment by schoolcoder** [comment]
Understood. Level 5. We've got 4 servers running, so either nothing will happen (everything's fixed), or one of them will have the problem within a few days.

**Comment by schoolcoder** [comment]
I parsed and analyzed the mutex activity in the Janus log file, and I believe I found the cause of the http hang.

There are a couple of items that may be related, so this is a long post. I'm no expert on Janus code, so please take any of my "should" statements as "schoolcoder thinks that..."

**Item 1: Mutex hanging (line 1839 in janus_http.c)**

The mutex lock on this line (function `janus_http_request_completed`) hangs. The call to `janus_mutex_lock` is shown here:
```
if(session) {
    janus_mutex_lock(&session->mutex);
    session->longpolls = g_list_remove(session->longpolls, ts);
    janus_mutex_unlock(&session->mutex);
    janus_refcount_decrease(&session->ref);
}
```

The session variable is set on line 1830:
```
janus_http_session *session = 
    (janus_http_session*)g_atomic_pointer_get(&request->longpoll);
```

However, `&request->longpoll` should be `request->longpoll` (no &). The extra & is causing `session` to contain junk, and then causing the mutex to hang.

My logfile contained hundreds of http and websocket transactions. The mutex above was referenced only once, and the attempt to lock was the last action of the http transport.

Furthermore, I believe the hang occurred when the http request was rejected. We have Janus client running http on 8088, proxied from Apache, but inadvertently had port 8088 open to the outside. Attackers were trying to connect using 'CONNECT' verb, which Janus correctly rejects in lines 1231-1234, as shown here:
```
if(strcasecmp(method, "GET") && strcasecmp(method, "POST") && strcasecmp(method, "OPTIONS")) {
    ret = janus_http_return_error(ts, 0, NULL, JANUS_ERROR_TRANSPORT_SPECIFIC, "Unsupported method %s", method);
    goto done;
}
```

At label `done` (line 1551), `janus_refcount_decrease(&ts->ref)` is called, but, in the case of a rejected request, there was no corresponding ` janus_refcount_increase(&ts->ref)`. It looks like that will cause `ts` to be freed, which could result in junk in the `janus_http_request_completed` function.


**Item 2: Duplicate mutex-lock in janus_websockets.c**

Our log showed the following lock activity (line numbers added):

```
1: [transports/janus_websockets.c:janus_websockets_send_message:933:lock] 0x7f07bc00c028
2: [transports/janus_websockets.c:janus_websockets_send_message:945:lock] 0x7f07ce3936a8
3: [transports/janus_websockets.c:janus_websockets_send_message:948:unlock] 0x7f07ce3936a8
4: [transports/janus_websockets.c:janus_websockets_common_callback:1134:lock] 0x7f07ce3936a8
5: [transports/janus_websockets.c:janus_websockets_common_callback:1146:unlock] 0x7f07ce3936a8
6: [transports/janus_websockets.c:janus_websockets_common_callback:1156:lock] 0x7f07bc00c028
7: [transports/janus_websockets.c:janus_websockets_send_message:956:unlock] 0x7f07bc00c028
8: [transports/janus_websockets.c:janus_websockets_common_callback:1204:unlock] 0x7f07bc00c028
9: [transports/janus_websockets.c:janus_websockets_common_callback:1156:lock] 0x7f07bc00c028
10: [transports/janus_websockets.c:janus_websockets_common_callback:1207:unlock] 0x7f07bc00c028
```

Mutex 0x7f07bc00c028 is the one of interest. It is locked on line 1, locked on 6, unlocked on 7, and unlocked on 8. (Then again locked on 9, unlocked on 10, but those are OK.)

This occurred when creating a room with videoroom plugin, using websockets transport.

This is not the cause of our http hang, but just wanted to point it out in case it's an issue.


**Item 3: free() should be g_free() (line 1946 in janus_http.c)**

**Item 4: Missing janus_refcount_decrease? (`janus_http_timeout` in janus_http.c)**

Function `janus_http_timeout` starts out like this:

```
void janus_http_timeout(janus_transport_session *ts, janus_http_session *session) {
    if(g_atomic_int_get(&ts->destroyed))
        return;
    janus_refcount_increase(&ts->ref);
    janus_http_msg *request = (janus_http_msg *)ts->transport_p;
    if(!g_atomic_int_compare_and_exchange(&request->timeout_flag, 1, 0)) {
        request->timeout = NULL;
        return;
    }
    ...
```

If the second `return` takes place, ts->ref count is left incremented (no call to `janus_refcount_decrease`). Not sure if this is a problem, but all other control paths in this function do call `janus_refcount_decrease`.

**Comment by atoppi** [comment]
Hi @schoolcoder ,
thanks for sharing your findings, replying inline:

> The session variable is set on line 1830:
> 
> ```
> janus_http_session *session = 
>     (janus_http_session*)g_atomic_pointer_get(&request->longpoll);
> ```
> 
> However, `&request->longpoll` should be `request->longpoll` (no &). The extra & is causing `session` to contain junk, and then causing the mutex to hang.

No, it's not. The extra & is needed ([glib ref](https://github.com/GNOME/glib/blob/7aa91a0289b7a6d15dc7bbe1c62d11f4c7b43974/glib/tests/atomic.c#L84), consider that `gpointer` is a `void*`). If this was wrong, every call to `janus_http_request_completed` would have crashed or hung Janus.

> My logfile contained hundreds of http and websocket transactions. The mutex above was referenced only once, and the attempt to lock was the last action of the http transport.

My guess is that `session` aka `request->longpoll` already contains junk at the moment the method starts.

> Furthermore, I believe the hang occurred when the http request was rejected. We have Janus client running http on 8088, proxied from Apache, but inadvertently had port 8088 open to the outside. Attackers were trying to connect using 'CONNECT' verb, which Janus correctly rejects in lines 1231-1234, as shown here:
> 
> ```
> if(strcasecmp(method, "GET") && strcasecmp(method, "POST") && strcasecmp(method, "OPTIONS")) {
>     ret = janus_http_return_error(ts, 0, NULL, JANUS_ERROR_TRANSPORT_SPECIFIC, "Unsupported method %s", method);
>     goto done;
> }
> ``` 
> At label `done` (line 1551), `janus_refcount_decrease(&ts->ref)` is called, but, in the case of a rejected request, there was no corresponding ` janus_refcount_increase(&ts->ref)`. It looks like that will cause `ts` to be freed, which could result in junk in the `janus_http_request_completed` function.

As far as I can tell, the transport session may be created (and its refcount initiated in `janus_transport_session_create`) or its reference may be increased in the _else_ branch at the beginning of  `janus_http_handler`.

```c
	if(ts == NULL) {
		firstround = 1;
		JANUS_LOG(LOG_DBG, "Got a HTTP %s request on %s...\n", method, url);
		JANUS_LOG(LOG_DBG, " ... Just parsing headers for now...\n");
		msg = g_malloc0(sizeof(janus_http_msg));
		msg->connection = connection;
		janus_refcount_init(&msg->ref, janus_http_msg_free);
		ts = janus_transport_session_create(msg, janus_http_msg_destroy);
		janus_mutex_lock(&messages_mutex);
		g_hash_table_insert(messages, ts, ts);
		janus_mutex_unlock(&messages_mutex);
		*ptr = ts;
		MHD_get_connection_values(connection, MHD_HEADER_KIND, &janus_http_headers, msg);
		ret = MHD_YES;
		/* Notify handlers about this new transport instance */
		if(notify_events && gateway->events_is_enabled()) {
			json_t *info = json_object();
			json_object_set_new(info, "event", json_string("request"));
			json_object_set_new(info, "admin_api", json_false());
			const union MHD_ConnectionInfo *conninfo = MHD_get_connection_info(connection, MHD_CONNECTION_INFO_CLIENT_ADDRESS);
			if(conninfo != NULL) {
				janus_network_address addr;
				janus_network_address_string_buffer addr_buf;
				if(janus_network_address_from_sockaddr((struct sockaddr *)conninfo->client_addr, &addr) == 0 &&
						janus_network_address_to_string_buffer(&addr, &addr_buf) == 0) {
					const char *ip = janus_network_address_string_from_buffer(&addr_buf);
					json_object_set_new(info, "ip", json_string(ip));
				}
				uint16_t port = janus_http_sockaddr_to_port((struct sockaddr *)conninfo->client_addr);
				json_object_set_new(info, "port", json_integer(port));
			}
			gateway->notify_event(&janus_http_transport, ts, info);
		}
	} else {
		JANUS_LOG(LOG_DBG, "Processing HTTP %s request on %s...\n", method, url);
		janus_refcount_increase(&ts->ref);
		msg = (janus_http_msg *)ts->transport_p;
	}
```

> 
> **Item 2: Duplicate mutex-lock in janus_websockets.c**
> 
> Our log showed the following lock activity (line numbers added):
> 
> ```
> 1: [transports/janus_websockets.c:janus_websockets_send_message:933:lock] 0x7f07bc00c028
> 2: [transports/janus_websockets.c:janus_websockets_send_message:945:lock] 0x7f07ce3936a8
> 3: [transports/janus_websockets.c:janus_websockets_send_message:948:unlock] 0x7f07ce3936a8
> 4: [transports/janus_websockets.c:janus_websockets_common_callback:1134:lock] 0x7f07ce3936a8
> 5: [transports/janus_websockets.c:janus_websockets_common_callback:1146:unlock] 0x7f07ce3936a8
> 6: [transports/janus_websockets.c:janus_websockets_common_callback:1156:lock] 0x7f07bc00c028
> 7: [transports/janus_websockets.c:janus_websockets_send_message:956:unlock] 0x7f07bc00c028
> 8: [transports/janus_websockets.c:janus_websockets_common_callback:1204:unlock] 0x7f07bc00c028
> 9: [transports/janus_websockets.c:janus_websockets_common_callback:1156:lock] 0x7f07bc00c028
> 10: [transports/janus_websockets.c:janus_websockets_common_callback:1207:unlock] 0x7f07bc00c028
> ```
> 
> Mutex 0x7f07bc00c028 is the one of interest. It is locked on line 1, locked on 6, unlocked on 7, and unlocked on 8. (Then again locked on 9, unlocked on 10, but those are OK.)
> 
> This occurred when creating a room with videoroom plugin, using websockets transport.
> 
> This is not the cause of our http hang, but just wanted to point it out in case it's an issue.

Unexpected indeed, let's hear what @lminiero thinks about this.

> 
> **Item 3: free() should be g_free() (line 1946 in janus_http.c)**

Right. Anyway there is a check in place for ptr nullity, so not an issue, just redundant code.

> 
> **Item 4: Missing janus_refcount_decrease? (`janus_http_timeout` in janus_http.c)**
> 
> Function `janus_http_timeout` starts out like this:
> 
> ```
> void janus_http_timeout(janus_transport_session *ts, janus_http_session *session) {
>     if(g_atomic_int_get(&ts->destroyed))
>         return;
>     janus_refcount_increase(&ts->ref);
>     janus_http_msg *request = (janus_http_msg *)ts->transport_p;
>     if(!g_atomic_int_compare_and_exchange(&request->timeout_flag, 1, 0)) {
>         request->timeout = NULL;
>         return;
>     }
>     ...
> ```
> 
> If the second `return` takes place, ts->ref count is left incremented (no call to `janus_refcount_decrease`). Not sure if this is a problem, but all other control paths in this function do call `janus_refcount_decrease`.

Will look into this.

**Comment by lminiero** [comment]
> > **Item 2: Duplicate mutex-lock in janus_websockets.c**
> > Our log showed the following lock activity (line numbers added):
> > ```
> > 1: [transports/janus_websockets.c:janus_websockets_send_message:933:lock] 0x7f07bc00c028
> > 2: [transports/janus_websockets.c:janus_websockets_send_message:945:lock] 0x7f07ce3936a8
> > 3: [transports/janus_websockets.c:janus_websockets_send_message:948:unlock] 0x7f07ce3936a8
> > 4: [transports/janus_websockets.c:janus_websockets_common_callback:1134:lock] 0x7f07ce3936a8
> > 5: [transports/janus_websockets.c:janus_websockets_common_callback:1146:unlock] 0x7f07ce3936a8
> > 6: [transports/janus_websockets.c:janus_websockets_common_callback:1156:lock] 0x7f07bc00c028
> > 7: [transports/janus_websockets.c:janus_websockets_send_message:956:unlock] 0x7f07bc00c028
> > 8: [transports/janus_websockets.c:janus_websockets_common_callback:1204:unlock] 0x7f07bc00c028
> > 9: [transports/janus_websockets.c:janus_websockets_common_callback:1156:lock] 0x7f07bc00c028
> > 10: [transports/janus_websockets.c:janus_websockets_common_callback:1207:unlock] 0x7f07bc00c028
> > ```
> > 
> > 
> > Mutex 0x7f07bc00c028 is the one of interest. It is locked on line 1, locked on 6, unlocked on 7, and unlocked on 8. (Then again locked on 9, unlocked on 10, but those are OK.)
> > This occurred when creating a room with videoroom plugin, using websockets transport.
> > This is not the cause of our http hang, but just wanted to point it out in case it's an issue.
> 
> Unexpected indeed, let's hear what @lminiero thinks about this.

There's nothing unexpected in this. We print attempts to lock _before_ we actually try to lock. This is needed to figure out deadlocks (e.g., if you see 5 locks for the same mutex and no unlock, then you have a problem). So two lock log lines for the same mutex are absolutely normal: it means the first got it, the second is waiting for it, and the two unlocks after that release it.

**Comment by atoppi** [comment]
Fixed Item 3 and 4 with [this commit](https://github.com/meetecho/janus-gateway/commit/4de7256c5fd6aceb7685bb851f5407ed772ed5cd).

We're closing the issue now, re-open in case the original problem should arise again.

**Comment by schoolcoder** [comment]
Thanks for the update, and your corrections to my original post.

We have been running our servers for 3 days now (port 8088 closed to the outside) without any trouble. The problematic session->mutex lock line has not been executed.

There's still a potential crash/memory-corruption problem lurking in janus_http_request_completed, though.

If session had garbage, then the line `janus_refcount_increase(&session->ref);` could cause memory corruption, even if the session->mutex lock doesn't cause a hang.

Maybe the session has already been destroyed by the time janus_http_request_completed is called? janus_http_request_completed does not check the session->destroyed flag....

**Comment by atoppi** [comment]
Hi @schoolcoder,
> Maybe the session has already been destroyed by the time janus_http_request_completed is called?

Yes that is true, I can not exclude a potential use-after-free there.

> janus_http_request_completed does not check the session->destroyed flag

That check would be quite useless though (if `session` has been totally freed).

Given that, should the crash happen again, please report back with a refcount debug and a libasan dump and we'll open again the issue.

**Comment by lminiero** [comment]
@schoolcoder you may want to get this commit: https://github.com/meetecho/janus-gateway/commit/436ccccabf100f9f454ccaf92afcf02bb7358aa0
It fixes an issue we were told about in #2118 which I suspect is the cause of your problem.
